From ae2c918e52c71aa5b9e17d0ab06ffa0e9e4fe689 Mon Sep 17 00:00:00 2001
From: maxwen <max.weninger@gmail.com>
Date: Fri, 23 Jun 2023 19:49:51 +0800
Subject: [PATCH 4/5] SettingsLib: Add deep sleep info to uptime preference

Add deep sleep ratio to "About Phone" -> "Uptime"
originally from Stefan

**DU Edits**
- Simplify the code used to show Deep sleep

Change-Id: I6949a66bae077b5132304448a7c8d7130665437a
---
 packages/SettingsLib/res/values/custom_strings.xml     |  7 +++++++
 .../deviceinfo/AbstractUptimePreferenceController.java | 10 +++++++++-
 2 files changed, 16 insertions(+), 1 deletion(-)
 create mode 100644 packages/SettingsLib/res/values/custom_strings.xml

diff --git a/packages/SettingsLib/res/values/custom_strings.xml b/packages/SettingsLib/res/values/custom_strings.xml
new file mode 100644
index 000000000000..a46e388da3c4
--- /dev/null
+++ b/packages/SettingsLib/res/values/custom_strings.xml
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+
+    <!-- Deep sleep information -->
+    <string name="status_deep_sleep">(Deep sleep: %1$d%2$s)</string>
+
+</resources>
diff --git a/packages/SettingsLib/src/com/android/settingslib/deviceinfo/AbstractUptimePreferenceController.java b/packages/SettingsLib/src/com/android/settingslib/deviceinfo/AbstractUptimePreferenceController.java
index 5f7226969699..5816578a63ea 100644
--- a/packages/SettingsLib/src/com/android/settingslib/deviceinfo/AbstractUptimePreferenceController.java
+++ b/packages/SettingsLib/src/com/android/settingslib/deviceinfo/AbstractUptimePreferenceController.java
@@ -26,6 +26,7 @@ import androidx.annotation.VisibleForTesting;
 import androidx.preference.Preference;
 import androidx.preference.PreferenceScreen;
 
+import com.android.settingslib.R;
 import com.android.settingslib.core.AbstractPreferenceController;
 import com.android.settingslib.core.lifecycle.Lifecycle;
 import com.android.settingslib.core.lifecycle.LifecycleObserver;
@@ -89,7 +90,14 @@ public abstract class AbstractUptimePreferenceController extends AbstractPrefere
     }
 
     private void updateTimes() {
-        mUptime.setSummary(DateUtils.formatElapsedTime(SystemClock.elapsedRealtime() / 1000));
+        float deepSleepRatio = Math.max((float) (SystemClock.elapsedRealtime() - SystemClock.uptimeMillis()), 0f)
+                / SystemClock.elapsedRealtime();
+        int deepSleepPercent = Math.round(deepSleepRatio * 100);
+
+        String summary = DateUtils.formatElapsedTime(SystemClock.elapsedRealtime() / 1000) +
+                " " + mContext.getString(R.string.status_deep_sleep, deepSleepPercent, "%");
+
+        mUptime.setSummary(summary.toString());
     }
 
     private static class MyHandler extends Handler {
-- 
2.34.1

